sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/ShardRecordsIterator.java

```diff
@@ -77,20 +77,28 @@ public ShardRecordsIterator(final ShardCheckpoint initialCheckpoint,
 
   private void readMoreIfNecessary() throws TransientKinesisException {
     if (data.isEmpty()) {
-      GetKinesisRecordsResult response;
+      GetKinesisRecordsResult response = fetchRecords();
+      data.addAll(filter.apply(response.getRecords(), checkpoint));
+    }
+  }
+
+  private GetKinesisRecordsResult fetchRecords() throws TransientKinesisException {
+    GetKinesisRecordsResult response = null;
+    do {
       try {
         response = kinesis.getRecords(shardIterator, checkpoint.getStreamName(),
             checkpoint.getShardId());
+        shardIterator = response.getNextShardIterator();
       } catch (ExpiredIteratorException e) {
         LOG.info("Refreshing expired iterator", e);
         shardIterator = checkpoint.getShardIterator(kinesis);
-        response = kinesis.getRecords(shardIterator, checkpoint.getStreamName(),
-            checkpoint.getShardId());
       }
-      LOG.debug("Fetched {} new records", response.getRecords().size());
-      shardIterator = response.getNextShardIterator();
-      data.addAll(filter.apply(response.getRecords(), checkpoint));
-    }
+    } while (response == null || gotEmptyResponseButIsBeforeEndOfTheStream(response));
+    return response;
+  }
+
+  private boolean gotEmptyResponseButIsBeforeEndOfTheStream(GetKinesisRecordsResult response) {
+    return response.getRecords().isEmpty() && response.getMillisBehindLatest() > 0;
```

Comment:
Hmm is this new exit condition related to watermark handling, or is it a bugfix? I'm not sure I understand the reason for the difference between old and new behavior.

Reply:
I actually answered that in my previous comment so just in short: Kinesis can return 0 records before reaching the end of the stream and that would cause problems with current watermark handling.